mixnet/mixnet.py

```diff
@@ -4,44 +4,59 @@
 from dataclasses import dataclass
 from typing import List
 
-from mixnet.fisheryates import FisherYates
 from mixnet.node import MixNode
 
 
 @dataclass
 class Mixnet:
-    mix_nodes: List[MixNode]
-
-    # Build a new topology deterministically using an entropy.
-    # The entropy is expected to be injected from outside.
-    #
-    # TODO: Implement constructing a new topology in advance to minimize the topology transition time.
-    #       https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#9a7f6089e210454bb11fe1c10fceff68
-    def build_topology(
-        self,
-        entropy: bytes,
-        n_layers: int,
-        n_nodes_per_layer: int,
-    ) -> MixnetTopology:
-        num_nodes = n_nodes_per_layer * n_layers
-        assert num_nodes < len(self.mix_nodes)
-
-        shuffled = FisherYates.shuffle(self.mix_nodes, entropy)
-        sampled = shuffled[:num_nodes]
-        layers = []
-        for l in range(n_layers):
-            start = l * n_nodes_per_layer
-            layer = sampled[start : start + n_nodes_per_layer]
-            layers.append(layer)
-        return MixnetTopology(layers)
-
-    def choose_mixnode(self) -> MixNode:
-        return random.choice(self.mix_nodes)
+    topology: MixnetTopology | None = None
+
+    def set_topology(self, topology: MixnetTopology) -> None:
+        """
+        Replace the old topology with the new topology received, and start establishing new network connections in background.
+
+        In real implementations, this method should be a long-running task, accepting topologies periodically.
+        Here in the spec, this method has been simplified as a setter, assuming the single-thread test environment.
+        """
+        self.topology = topology
+        self.establish_connections(topology)
+
+    def get_topology(self) -> MixnetTopology:
+        if self.topology is None:
+            raise RuntimeError("topology is not set yet")
+        return self.topology
+
+    @staticmethod
+    def establish_connections(_: MixnetTopology) -> None:
```

Comment:
nitpic: shouldn't this be a method with self instead?

Reply:
Absolutely. Thank you. This is a leftover from when I tried to make it async with some dummy implementation. But now, as you can said, I think it's enough to have a comment here instead of something complicated. I just modified it to a method with `self`: d145107